src/main/java/at/ac/tuwien/kr/alpha/common/atoms/BuiltinAtom.java

```diff
@@ -0,0 +1,222 @@
+package at.ac.tuwien.kr.alpha.common.atoms;
+
+import at.ac.tuwien.kr.alpha.common.BinaryOperator;
+import at.ac.tuwien.kr.alpha.common.predicates.FixedInterpretationPredicate;
+import at.ac.tuwien.kr.alpha.common.predicates.Predicate;
+import at.ac.tuwien.kr.alpha.common.terms.ArithmeticTerm;
+import at.ac.tuwien.kr.alpha.common.terms.ConstantTerm;
+import at.ac.tuwien.kr.alpha.common.terms.Term;
+import at.ac.tuwien.kr.alpha.common.terms.VariableTerm;
+import at.ac.tuwien.kr.alpha.grounder.Substitution;
+
+import java.util.*;
+
+/**
+ * Represents a builtin atom according to the standard.
+ * Copyright (c) 2017, the Alpha Team.
+ */
+public class BuiltinAtom extends FixedInterpretationAtom {
```

Comment:
I think the main problem is that the terms of such fixed interpretation predicates still have an influence on the grounding besides just yielding other substitutions. In the case of builtin equality they also influence the order in which a rule may be grounded and they influence whether a rule is safe or not by sometimes yielding bindings for variables.

Examples: For `X = a` the equality binds the variable `X`, while for `f(X) = f(a)` it binds no variable and for `X +1 = Y` it binds `Y`.

So the predicate alone does not have enough information to determine the `getBindingVariables`. This was my main motivation to put the whole treatment into the atom instead of the predicate. On the other hand, of course, the evaluation at the predicate is now mostly useless since only the one at the atom is used and only in case of the external, it relays to the predicate. So I guess, we could also get rid of the evaluation at the predicate in my approach?

But I also start to see that the other approach (splitting variable binding from the evaluation, then move the evaluation away from the atom) might work. However, I see no clear picture for that at the moment. I am not against trying this direction. What benefits besides `Predicate` just being a symbol could we get?

Reply:
I mainly wanted to have this distinction since externals get confusing if we do not split predicate and its interpretation. For example, we have no nice way of representing the arity of both input and output terms.